File: day2.hs
import AOC

main = aoc "day2" part1 part2

-- | Move instruction with tuple of (forward distance, down distance)
-- Negative distance forward means backwards, and negative down means move up
data Instruction = Move Int Int deriving (Show, Eq)

instance Read Instruction where
    readsPrec _ ('u':'p':' ':n) = [(Move 0 ((-1) * read n), "")]
    readsPrec _ ('d':'o':'w':'n':' ':n) = [(Move 0 (read n), "")]
    readsPrec _ ('f':'o':'r':'w':'a':'r':'d':' ':n) = [(Move (read n) 0, "")]

instance Semigroup Instruction where
    (Move a b) <> (Move c d) = Move (a + c) (b + d)

instance Monoid Instruction where
    mempty = Move 0 0

part1 :: [String] -> Int
part1 input = a * b
  where
    Move a b = mconcat $ map read input


-- For the second type of instruction, we need to maintain state of the aim
-- direction, so a Monoid won't really work.
data Instruction2 = Forward Int | ShiftAim Int deriving (Show, Eq)

instance Read Instruction2 where
    readsPrec _ ('d':'o':'w':'n':' ':n) = [(ShiftAim (read n), "")]
    readsPrec _ ('u':'p':' ':n) = [(ShiftAim ((-1) * read n), "")]
    readsPrec _ ('f':'o':'r':'w':'a':'r':'d':' ':n) = [(Forward (read n), "")]

part2 :: [String] -> Int
part2 input = uncurry (*) $ execInstr2 (0, 0) 0 $ map read input

execInstr2 :: (Int, Int) -> Int -> [Instruction2] -> (Int, Int)
execInstr2 (x, y) aim [] = (x, y)
execInstr2 (x, y) aim (Forward n : xs) = execInstr2 (x + n, y + (aim * n)) aim xs
execInstr2 (x, y) aim (ShiftAim n : xs) = execInstr2 (x, y) (aim + n) xs

File: src/day2.ml
open Base
open AOC

let result (a, b) = a * b

let (|-) x f = f x; x

type instruction1 = Move of int * int

let parse_instructions1 input =
  let [instr; amount] = String.split_on_chars ~on:[' '] input in
  match instr with
  | "up"      -> Move (0, (-1) * Int.of_string amount)
  | "down"    -> Move (0, Int.of_string amount)
  | "forward" -> Move (Int.of_string amount, 0)
  | _         -> Move (0, 0)

let rec exec1 = function
| []                  -> (0, 0)
| (Move (a, b) :: xs) -> let (x, y) = exec1 xs in (x + a, y + b)

let part1 input =
  input
  |> List.map ~f:parse_instructions1
  |> exec1
  |> result
  |> Int.to_string

type instruction2
  = Forward of int
  | ShiftAim of int

let parse_instructions2 input =
  let [instr; amount] = String.split_on_chars ~on:[' '] input in
  match instr with
  | "up"      -> ShiftAim ((-1) * Int.of_string amount)
  | "down"    -> ShiftAim (Int.of_string amount)
  | "forward" -> Forward (Int.of_string amount)

let rec exec2 (x, y) aim = function
| []                 -> (x, y)
| (Forward n :: xs)  -> exec2 (x + n, y + (aim * n)) aim xs
| (ShiftAim n :: xs) -> exec2 (x, y) (aim + n) xs

let part2 input =
  input
  |> List.map ~f:parse_instructions2
  |> exec2 (0, 0) 0
  |> result
  |> Int.to_string

let () = aoc "day2" part1 part2
